### ISSUE TITLE
flet pack additional options (-D, --onedir)

### ISSUE THREAD

**Issue by DavMelchi** [flet-dev/flet#804]
Hi,
Currently `**flet pack** ` allows only one-file bundled
It would be nice if we can have "**-D, --onedir**" option for flet
Sometimes when the .exe size is big, it takes longer to start with one-file bundled than **one-folder bundle app**.
Thanks!

**Comment by FeodorFitsner** [comment]
Will take a look. I remember that were some issues with macOS bundle and `--onedir`, but Windows should be OK.

**Comment by DavMelchi** [comment]
Thank You @FeodorFitsner

**Comment by ndonkoHenri** [comment]
Please, Is it possible to have an option for the `--onefile` ? (to obtain a single-file executable)

**Comment by DavMelchi** [comment]
> Please, Is it possible to have an option for the `--onefile` ? (to obtain a single-file executable)

Did you try it ?  Currently, it is already like this by default. You'll get a single executable with flet pack.

**Comment by ndonkoHenri** [comment]
I tried it, had errors and gave up quite easily. Let me try again now.

**Comment by ndonkoHenri** [comment]
Just from trying again. still had errors, but from what i see, @DavMelchi , seems like what I meant is not the default. What flet pack does is create new folders (dist, build) and files (main.spec).
The to share this file, we will need to share up all the generated bundle(files/folders).

What I meant was making it possible for flet pack to generated only one file: the single executable file. In pyinstaller this is done with the `--onefile` optional command (https://pyinstaller.org/en/stable/usage.html#cmdoption-F)
Because it's just one file, it is easy to share or upload to github releases. It makes the users happy to only have to see on executable file.

**Comment by DavMelchi** [comment]
@ndonkoHenri  At the end you will have single exe in dist folder. You can move it wherever you want and delete the build folder and .spec file. They no longer matter. I tried and it works for me

Sharing your exe is not a problem whether it's single exe or folder with` Inno Setup`. It's even more professional if you want your program to be installable and not just portable.

**Comment by AbortLarboard** [comment]
the _--onedir_ is quite nice because startup times are a bit faster. No need to unpack everything in a __MEI_ folder in in _:\Users\user\AppData\Local\Temp\ everytime you start the app._

if you got something like a config file e.g. sqlite db bundled you still will need to create/bundle it in the directory of the final.exe since the _MEI folders will be different everytime you start the flet app.

**Comment by jazzy283** [comment]
Note to add the ability to deal with multiple --add-data arguments to the "flet pack" wrapper

**Comment by thearyadev** [comment]
feature request: 

Additional pyinstaller arguments to be appended to the list of arguments passed to `flet pack`
![image](https://user-images.githubusercontent.com/87589047/212592933-40ec5049-604b-4eda-96fc-d2f4a582cf63.png)

**Comment by FeodorFitsner** [comment]
All there `--onedir` (`--onefile` is default), `--add-data` and `--hidden-import` arguments have been added and can be tested in a pre-release version of Flet (`pip install flet --pre`). Let me know how that worked.

**Comment by DavMelchi** [comment]
Hi @FeodorFitsner ,
I tried to run my app on the pre release but it doesn't work properly. (Buttons with asset images that don't appear properly and error messages that I don't understand ...) I'm a little busy at the moment to try figure out where the problem is coming from, but if I find it, I'll get back to you.

![image](https://user-images.githubusercontent.com/116588982/212877721-0d64a4eb-a466-43ed-9295-208318319b13.png)

**Comment by FeodorFitsner** [comment]
@DavMelchi Thanks for trying! This is why I hesitate to release a build with that yet - I suspect there might be some edge cases. It would be great if you could share the code reproducing the issue, so I can investigate.

**Comment by DavMelchi** [comment]
Hi @FeodorFitsner 
Please find below the link of the code to reproduce the error.
Just run the Main.py to get the result.
I used _RoundedRectangleBorder, BorderSide_. Don't forget to adapt the imports if you test on the latest version please

https://github.com/DavMelchi/Flet_PreTest.git

Result with :  0.4.0.dev1000

![image](https://user-images.githubusercontent.com/116588982/212987852-0906b226-e7ac-46db-bf97-ace0cd9e9988.png)

Result with : 0.4.0.dev1065

![image](https://user-images.githubusercontent.com/116588982/212988403-cd4f5e53-3d4c-45d2-bedf-9a2a57d6ad50.png)

![image](https://user-images.githubusercontent.com/116588982/212989411-cbeb2f84-d0cb-4711-9f07-27f33b6c44a2.png)

**Comment by FeodorFitsner** [comment]
Thanks, will take a look!

**Comment by FeodorFitsner** [comment]
@DavMelchi I didn't catch a JSON decoding bug, but fixed a loading of assets.

For desktop apps a path in `src` property could be one of the following:
  * A path relative to `assets` directory, with or without starting slash, for example: `/image.png` or `image.png`. The name of artifact dir should not be included.
  * An absolute path within a computer file system, e.g. `C:\projects\app\assets\image.png` or `/Users/john/images/picture.png`.
  * A full URL, e.g. `https://mysite.com/images/pic.png`.

So, in your app it should be:

```python
            MyElevetatedButton(
                f"home.svg",
                "Button1",
                ButtonClick,
                60,
                30,
            ),
```

Upgrade Flet to `0.4.0.dev1066`.

Are you still having JSON decoder issue?

**Comment by DavMelchi** [comment]
Thanks a lot @FeodorFitsner ,
This error is corrected with version `0.4.0.dev1066` and the path instructions as you indicated.

And for our main topic, I tried `--onedir` and it works perfectly on my side.

But I have a new json error in another part of my app with Mysql-DataTable-Filter etc... and at the moment I don't know where it is coming from. I will continue to search and and provide a sample code to reproduce if I find it.

![image](https://user-images.githubusercontent.com/116588982/213239388-38be6744-34e0-411d-9f5f-5f7f553c71a1.png)


(I don't have this error with version 0.4.0.dev1000)
Looks like my codes are problematic and the new version doesn't like bullshit :)

**Comment by FeodorFitsner** [comment]
@DavMelchi I don't think it's something wrong with your program. Could you please enable debug logging by adding the following at the beginning of your program and paste the output here (as a text please, not a screenshot):

```python
import logging
logging.basicConfig(level=logging.DEBUG)
```

**Comment by DavMelchi** [comment]
DEBUG:root:Sent: 19995
DEBUG:root:Message size: 98
DEBUG:root:_on_message: {"action":"pageEventFromWeb","payload":{"eventTarget":"_3827","eventName":"focus","eventData":""}}
INFO:root:page.on_event: _3827 focus
DEBUG:root:Message size: 97
DEBUG:root:_on_message: {"action":"pageEventFromWeb","payload":{"eventTarget":"_3827","eventName":"blur","eventData":""}}
INFO:root:page.on_event: _3827 blur
DEBUG:root:Message size: 88
DEBUG:root:_on_message: {"action":"updateControlProps","payload":{"props":[{"i":"_3827","value":"1ere année"}]}
DEBUG:root:Cloding connection
Exception in thread Thread-1 (__connection_loop):
Traceback (most recent call last):
  File "C:\Program Files\Python311\Lib\threading.py", line 1038, in _bootstrap_inner
    self.run()
  File "C:\Program Files\Python311\Lib\threading.py", line 975, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Users\HP\Documents\PROJET FLUTTER\FLET\STM\.venv\Lib\site-packages\flet\sync_local_socket_connection.py", line 146, in __connection_loop
    self.__on_message(message.decode("utf-8"))
  File "C:\Users\HP\Documents\PROJET FLUTTER\FLET\STM\.venv\Lib\site-packages\flet\sync_local_socket_connection.py", line 67, in __on_message
    msg_dict = json.loads(data)
               ^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\json\__init__.py", line 346, in loads
    return _default_decoder.decode(s)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\json\decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\json\decoder.py", line 353, in raw_decode
    obj, end = self.scan_once(s, idx)
               ^^^^^^^^^^^^^^^^^^^^^^
json.decoder.JSONDecodeError: Expecting ',' delimiter: line 1 column 88 (char 87)

**Comment by FeodorFitsner** [comment]
Thank you! I've been able to produce this unicode-related issue:

![image](https://user-images.githubusercontent.com/5041459/213267715-55ab866b-23f0-45f0-88f6-980c3c67c1c2.png)

**Comment by DavMelchi** [comment]
Haha... My head is hot here, and I couldn't even figure out where the problem was. 😂

**Comment by FeodorFitsner** [comment]
Please install https://pypi.org/project/flet/0.4.0.dev1070/ with encoding bugfix and let me know how it worked.

**Comment by DavMelchi** [comment]
Thank you @FeodorFitsner  
Everything is ok on my side now.
For flet pack , `--onedir` and multiple` --add-data` are working well also for me.
let's wait @thearyadev  feedback  for `--hidden-import` according to his use case.
Thanks

**Comment by jazzy283** [comment]
New options work perfectly for me. Thanks!!
